core/src/main/java/io/questdb/griffin/engine/functions/math/WidthBucketFunctionFactory.java

```diff
@@ -0,0 +1,129 @@
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.math;
+
+import org.jetbrains.annotations.Nullable;
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.cairo.sql.SymbolTableSource;
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.IntFunction;
+import io.questdb.griffin.engine.functions.QuarternaryFunction;
+import io.questdb.std.IntList;
+import io.questdb.std.Numbers;
+import io.questdb.std.ObjList;
+
+public class WidthBucketFunctionFactory implements FunctionFactory {
+    @Override
+    public String getSignature() {
+        return "width_bucket(DDDI)";
+    }
+
+    @Override
+    public Function newInstance(int position, ObjList<Function> args, IntList argPositions,
+            CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        return new WidthBucketFunction(args.getQuick(0), args.getQuick(1), args.getQuick(2), args.getQuick(3));
+    }
+
+    private static abstract class WidthBucketFunction extends IntFunction implements QuarternaryFunction {
+        private final Function operandFunc;
+        private final Function lowFunc;
+        private final Function highFunc;
+        private final Function countFunc;
+
+        public WidthBucketFunction(Function operandFunc, Function lowFunc, Function highFunc,
+                Function countFunc) {
+            this.operandFunc = operandFunc;
+            this.lowFunc = lowFunc;
+            this.highFunc = highFunc;
+            this.countFunc = countFunc;
+        }
+
+        @Override
+        public Function getLeftEnd() {
+            return operandFunc;
+        }
+
+        @Override
+        public Function getCenterLeft() {
+            return lowFunc;
+        }
+
+        @Override
+        public Function getCenterRight() {
+            return highFunc;
+        }
+
+        @Override
+        public Function getRightEnd() {
+            return countFunc;
+        }
+
+        @Override
+        public void init(SymbolTableSource symbolTableSource, SqlExecutionContext executionContext)
+                throws SqlException {
+            QuarternaryFunction.super.init(symbolTableSource, executionContext);
+            if (countFunc.isRuntimeConstant()) {
```

Comment:
We need to move this and the later checks to `WidthBucketFunctionFactory#newInstance()` and add `isConstant()` condition. Otherwise, users will be able to pass arbitrary functions as the corresponding arguments and they will bypass the validation.

We also need to the error checks with tests. See this test as an example:
https://github.com/questdb/questdb/blob/a17c85567e50b419783774484fba6bbd601735f5/core/src/test/java/io/questdb/griffin/engine/functions/str/SplitPartFunctionFactoryTest.java#L37-L45